bookkeeper-server/src/test/java/org/apache/bookkeeper/client/api/BookKeeperApiTest.java

```diff
@@ -0,0 +1,300 @@
+/*
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+package org.apache.bookkeeper.client.api;
+
+import static org.apache.bookkeeper.common.concurrent.FutureUtils.result;
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+
+import io.netty.buffer.Unpooled;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import org.apache.bookkeeper.client.BKException;
+import org.apache.bookkeeper.client.BKException.BKDigestMatchException;
+import org.apache.bookkeeper.client.BKException.BKDuplicateEntryIdException;
+import org.apache.bookkeeper.client.BKException.BKLedgerFencedException;
+import org.apache.bookkeeper.client.BKException.BKNoSuchLedgerExistsException;
+import org.apache.bookkeeper.client.BKException.BKUnauthorizedAccessException;
+import org.apache.bookkeeper.client.MockBookKeeperTestCase;
+import org.apache.bookkeeper.proto.BookieProtocol;
+import org.junit.Test;
+
+/**
+ * Unit tests of classes in this package
+ */
+public class BookKeeperApiTest extends MockBookKeeperTestCase {
+
+    @Test
+    public void testWriteHandle() throws Exception {
+        Map<String, byte[]> customMetadata = new HashMap<>();
+        customMetadata.put("test", "test".getBytes(StandardCharsets.UTF_8));
+        try (WriteHandle writer
+            = result(
+                newCreateLedgerOp()
+                    .withAckQuorumSize(1)
+                    .withWriteQuorumSize(2)
+                    .withEnsembleSize(3)
+                    .withDigestType(DigestType.MAC)
+                    .withCustomMetadata(customMetadata)
+                    .withPassword("password".getBytes(StandardCharsets.UTF_8))
+                    .execute());) {
+
+            byte[] data = "foo".getBytes(StandardCharsets.UTF_8);
+            writer.append(ByteBuffer.wrap(data)).get();
```

Comment:
1) what do you want to test between line 65 to line 69?  why not write it in a loop? what is the difference between line 67 and other lines?
2) you might consider verifying the mock BookieClient on how many times it is called?